bqskit/compiler/passes/partitioning/scan.py

```diff
@@ -180,59 +265,41 @@ def run(self, circuit: Circuit, data: dict[str, Any]) -> None:
-                    }),
-                )
-
-                in_qudits = list(q for q in qudit_group)
-                stopped_cycles = {q: num_cycles for q in qudit_group}
-                score = 0
-
-                for cycle, op in ops_and_cycles:
-                    if len(op.location.union(in_qudits)) != len(in_qudits):
-                        for qudit_index in op.location.intersection(in_qudits):
-                            stopped_cycles[qudit_index] = cycle
-                            in_qudits.remove(qudit_index)
-                    else:
-                        if len(op.location) > 1:
-                            score += self.multi_gate_score
-                        else:
-                            score += self.single_gate_score
-
-                    if len(in_qudits) == 0:
-                        break
-
-                if best_score is None or score > best_score:
-                    best_score = score
-                    best_region = CircuitRegion({
-                        qudit: (
-                            divider[qudit],
-                            # Might have errors if below is removed
-                            stopped_cycles[qudit] - 1,
-                        )
-                        for qudit in qudit_group
-                        # This statement
-                        # if stopped_cycles[qudit] - 1 >= divider[qudit]
-                    })
+            # Heap partitioning
+            # Get best (up to date) block
+            while len(score_heap) > 0:
+                (best_score, group_index) = heapq.heappop(score_heap)
+                best_score *= -1
+                if best_score == block_list[group_index].score:
+                    best_region = block_list[group_index].region
+                    break
 
             if best_score is None or best_region is None:
                 raise RuntimeError('No valid block found.')
 
             _logger.info('Found block with score: %d.' % (best_score))
             regions.append(best_region)
 
-            # Update divider
+            # Update divider, find blocks to rescore
+            rescore_set = []
             for qudit_index in best_region:
                 divider[qudit_index] = best_region[qudit_index].upper + 1
+                rescore_set.extend(member_dict[qudit_index])
+
+            rescore_set = list(set(rescore_set))
+
+            for group_index in rescore_set:
+                # Find the new region and score
+                (new_region, new_score) = self._form_region_and_score(
+                    block_list[group_index].qudit_group,
+                    divider,
+                    circuit,
+                )
+                # Update the block_list
+                block_list[group_index].region = new_region
+                block_list[group_index].score = new_score
+                # push to heap
+                heapq.heappush(score_heap, (-1 * new_score, group_index))
```

Comment:
Multiple inserts into a heap can be made more efficient by first building a heap of the elements to insert then merging the two heaps. You can use `heapq.merge` to the do the merging.

See [here](https://cs.stackexchange.com/questions/105086/an-algorithm-to-efficiently-insert-a-list-of-elements-into-a-binary-heap-bulk) for more info.

Reply:
Thanks for pointing that out, I've switched to this.